README.md
###Please read
**From release 1.3, code folding will be integrated into Brackets (although without custom region folding). This repository is still active for users who are running Brackets versions less than 1.3. If you are running 1.3 and would like custom region code folding you can find it [here](https://github.com/thehogfather/brackets-custom-region-code-folding).**

#Code folding for Brackets
A code folding extension for [Brackets](https://github.com/adobe/brackets/) based on [CodeMirror's folding addon](http://codemirror.net/demo/folding.html).
Peforms code folding based on brace ({}, []) matching and multiline comments for javascript, json, css, php and less files. Also has some support for html and xml files based on tag matching.

![Alt text](./screenshots/folded-example.png?raw=true "Folded Example")

### Features
1. Define custom regions in any language recognised in Brackets by creating a line comment whose first word is **region** to designate the beginning of a region, and another line comment whose first word is **endregion** to designate the end of a region

2. Collapse or expand all foldable child regions by holding down the Alt-key while clicking on fold triangles (as found in OSX Finder).

3. Fold states are persisted to preferences file.

### How to install
Navigate to **Brackets > File > Install Extension** and paste url https://github.com/thehogfather/brackets-code-folding

####Or manually
1. [Download](https://github.com/thehogfather/brackets-code-folding/archive/master.zip)
2. Unzip in **user** folder in **Brackets > Help > Show Extensions Folder**
3. Restart or Reload Brackets

### Keyboard shortcuts
The following are the default keyboard shortcut keys for the code-folding extension.

    Ctrl-Alt-C Collapse code region at current cursor position
    Ctrl-Alt-X Expand code region at current cursor position
    Alt-1 Collapse all code regions in current editor
    Shift-Alt-1 Expand all code regions in current editor


##Known Issues
#### Conflicting keyboard shortcuts
**If any of the shortcut keys above conflict with your keyboard or you would prefer to use another shortcut key for any reason, you can [remap the keys as described here](https://github.com/adobe/brackets/wiki/User-Key-Bindings). **

The following are CommandIds for remmaping keys

    codefolding.collapse.all                **collapses all regions**
    codefolding.collapse                    **collapses current region**
    codefolding.expand                      **expands current region**
    codefolding.expand.all                  **expands all code regions in file**
    codefolding.collapse.customregions      **collapses all custom regions in the file **
    codefolding.settings                    **show code folding settings dialog**

#### EdgeCode support
If you are running edge code, the latest compatible version is [v0.2.7](https://github.com/thehogfather/brackets-code-folding/tree/v0.2.7). You can download and install from [this url](https://github.com/thehogfather/brackets-code-folding/tree/v0.2.7).

### License
MIT-licensed.

### Credits
CodeMirror folding addon.

### Compatibility
Brackets Sprint 38 or later.


CHANGELOG.md
#Changelog
##0.3.5
* Addresses #134 where the beginning of lines in files are occluded by the fold-gutter.
* Addresses #133 where disabling code-folding for the settings dialog makes the settings inaccessible on reload
##0.3.4
* Updated for upcoming integration of code-folding in brackets master. Extension is no longer compatible with brackets >= 1.4.
##0.3.2
* Addresses #122 Cant open fold after tab or space on folded line
##0.3.1
* Addresses #119 Possible CodeMirror conflict with brackets-git
##0.3.0
* Addresses #114 gaps no longer appear in fold gutter (especially in custom themes)
* Addresses #116 external modification to files no longer corrupt the fold states in the document

##0.2.26
* Addresses #110: marker for folded region now styled using only css.
##0.2.25
* Added max fold levels to the settings dialog. This improves performance when folding all regions in a document by restricting the level of nesting to a specified number.
* Updated Russian translation
* Updated Finnish translation
* Added portuguese (Brasil) translation [rafaelcastrocouto](https://github.com/rafaelcastrocouto)
* Added Dutch translation [BBosman](https://github.com/BBosman)
##0.2.24
* Addresses issue #103 - code folding plugin no longer overwrites other gutter elements
* Added Finish translation [valtlait](https://github.com/valtlait)
##0.2.23
* Added options to fade fold buttons [vadim-kudr](https://github.com/vadim-kudr)
##0.2.22
* Patch for issue #88 - any space character now allowed before a region fold
* Added Japanese translation [lclzd](https://github.com/lclzd)
##0.2.21
* Added menu item for collapsing all custom regions in a document (feature #85)
* Updated Russian translation
##0.2.20
* Improved performance when opening minified files with region folding enabled
* Added French translation [rainje](https://github.com/rainje)
* Added Russian translation [disshishkov](https://github.com/disshishkov)
##0.2.19
* Added Settings dialog for configuring the extension

##0.2.18
* Added Galician translation [ivarcia ](https://github.com/ivarcia)
##0.2.17
* BUGFIX - fixed issue #78 - region fold causes brackets IDE to freeze when opening certain minified javascript files.

##0.2.16
* Added Italian translation [Denisov21](https://github.com/Denisov21)
* Update to Spanish translation [JJBocanegra](https://github.com/JJBocanegra)

##0.2.15
* Refactored and simplified region fold so it works for any language recognised in codeMirror
* Added custom region folding [keleko34](https://github.com/keleko34)
* Added Spanish translation [QuijoteShin](https://github.com/QuijoteShin)

##0.2.14
* BUGFIX - fixed issue #67 regarding nested fold regions. When a region becomes invalid it is now automatically expanded.

##0.2.13
* BUGFIX - fixed issue #66 where tags (in html/xml) wont fold if the tag is selected

##0.2.12
* BUGFIX - addresses issue #64 where extension crashed when loading preferences

##0.2.11
* Updated package.json to reflect correct minimum brackets version required to run extension

##0.2.10
* BUGFIX - fixed issue #61 where fold state was not correctly persisted
* Simplified the datastructure for storing foldstates in preference file

##0.2.9
* BUGFIX - fixed issue #57

##0.2.8
* Load code folding addons directly from brackets codemirror addon

##0.2.7
* BUGFIX - fixed issue #51 where extension was crashing for edgecode

##0.2.6
* BUGFIX - fixed issue #50
* Upgraded preference persistence to the new Brackets API

##0.2.5
* BUGFIX - fixed issue #49 where collapse all did not include first foldable item

##0.2.4
* Updated package.json to include title
* BUGFIX - fixed issue #44 [Mafio](https://github.com/Mafio)

##0.2.3
* Added internationalisation [Mafio](https://github.com/Mafio)
* Updated keyboard shortcuts

##0.2.2
* Several improvements to indent fold addressing issue #40

##0.2.1
* made indent the default fold addon for text files to address issue #39

##0.2.0
* added support for ruby files

##0.1.8
* Improved latex code folding
* BUGFIX - fixed issue #37 where multiline comments no longer had fold markers

##0.1.7
* Added support for folding latex documents

##0.1.6
* Change base key for shortcuts to letters in order to reduce chances of conflicts with international keyboards addresses issues #34, #35 and #36
##0.1.4
* Added menu items for collapsing and expanding code regions under view menu (addresses issue #33)
* Removed keybinding (ctrl-alt-+) for expanding current code region and left (ctrl-alt-=) as per issue #34

##0.1.3
* Added instruction to README to address issue #34 where keyboard shortcuts might conflict with typing non alphabetic characters on some non-English keyboards.

##0.1.2
* Addresses issue #32 where line fold state is not being correctly persisted after modifying document. Consequently restoring line folds results in wrong folded regions.

##0.1.1
* Addresses issue #30 where code folding is broken after using Ctrl-Alt-- to fold current code region.
* Fixed minor issue where code range opened and closed on the same line is still marked as folded when using shortcuts


main.js
/*
* Copyright (c) 2013 Patrick Oladimeji. All rights reserved.
*
* Permission is hereby granted, free of charge, to any person obtaining a
* copy of this software and associated documentation files (the "Software"),
* to deal in the Software without restriction, including without limitation
* the rights to use, copy, modify, merge, publish, distribute, sublicense,
* and/or sell copies of the Software, and to permit persons to whom the
* Software is furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
* FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
* DEALINGS IN THE SOFTWARE.
*
*/
/**
 * Code folding extension for brackets
 * @author Patrick Oladimeji
 * @date 10/24/13 9:35:26 AM
 */
/*jslint vars: true, plusplus: true, devel: true, nomen: true, indent: 4, maxerr: 50 */
/*global define, require, $, brackets*/

require.config({
    paths: {
        "text" : "lib/text",
        "i18n" : "lib/i18n"
    },
    locale: brackets.getLocale()
});

define(function (require, exports, module) {
    "use strict";
    var CodeMirror              = brackets.getModule("thirdparty/CodeMirror2/lib/codemirror"),
        Strings                 = require("strings"),
        AppInit                 = brackets.getModule("utils/AppInit"),
        CommandManager          = brackets.getModule("command/CommandManager"),
        DocumentManager         = brackets.getModule("document/DocumentManager"),
        EditorManager           = brackets.getModule("editor/EditorManager"),
        ProjectManager          = brackets.getModule("project/ProjectManager"),
        KeyBindingManager       = brackets.getModule("command/KeyBindingManager"),
        ExtensionUtils          = brackets.getModule("utils/ExtensionUtils"),
        Menus					= brackets.getModule("command/Menus"),
        prefs                   = require("./Prefs"),
        COLLAPSE_ALL            = "codefolding.collapse.all",
        COLLAPSE                = "codefolding.collapse",
        EXPAND                  = "codefolding.expand",
        EXPAND_ALL              = "codefolding.expand.all",
        GUTTER_NAME             = "CodeMirror-foldgutter",
        COLLAPSE_CUSTOM_REGIONS = "codefolding.collapse.customregions",
        CODE_FOLDING_SETTINGS	= "codefolding.settings",
        SettingsDialog          = require("SettingsDialog"),
        collapseKey             = "Ctrl-Alt-[",
        expandKey               = "Ctrl-Alt-]",
        collapseAllKey          = "Alt-1",
        expandAllKey            = "Shift-Alt-1";


    //load code mirror addons
    brackets.getModule(["thirdparty/CodeMirror2/addon/fold/brace-fold"]);
    brackets.getModule(["thirdparty/CodeMirror2/addon/fold/comment-fold"]);
    brackets.getModule(["thirdparty/CodeMirror2/addon/fold/markdown-fold"]);

    //still using slightly modified versions of the foldcode.js and foldgutter.js since we
    //need to modify the gutter click handler to take care of some collapse and expand features
    //e.g. collapsing all children when 'alt' key is pressed
    var foldGutter              = require("foldhelpers/foldgutter"),
        foldCode                = require("foldhelpers/foldcode");

    var indentFold              = require("foldhelpers/indentFold"),
        regionFold              = require("foldhelpers/region-fold");

    /**
        Restores the linefolds in the editor using values fetched from the preference store
        Checks the document to ensure that changes have not been made (e.g., in a different editor)
        to invalidate the saved line folds.
        @param {Editor} editor  the editor whose saved line folds should be restored
        */
    function restoreLineFolds(editor) {
        var saveFolds = prefs.getSetting("saveFoldStates");
        if (!editor || !saveFolds) {
            return;
        }
        var cm = editor._codeMirror;
        var path = editor.document.file.fullPath;
        var folds = cm._lineFolds || prefs.getFolds(path);
        cm._lineFolds = cm.getValidFolds(folds);
        prefs.setFolds(path, cm._lineFolds);
        Object.keys(cm._lineFolds).forEach(function (line) {
            cm.foldCode(Number(line));
        });
    }

    /**
      * Saves the line folds in the editor using the preference storage
      * @param {Editor} editor the editor whose line folds should be saved
      */
    function saveLineFolds(editor) {
        var saveFolds = prefs.getSetting("saveFoldStates");
        if (!editor || !saveFolds) {
            return;
        }
        var folds = editor._codeMirror._lineFolds || {};
        var path = editor.document.file.fullPath;
        if (Object.keys(folds).length) {
            prefs.setFolds(path, folds);
        } else {
            prefs.setFolds(path, undefined);
        }
    }

    /**
      * Event handler for gutter click. Manages folding and unfolding code regions. If the Alt key
      * is pressed while clicking the fold gutter, child code fragments are also folded/unfolded
      * up to a level defined in the `maxFoldLevel' preference.
      * @param {!CodeMirror} cm the CodeMirror object
      * @param {number} line the line number for the clicked gutter
      * @param {string} gutter the name of the gutter element clicked
      * @param {!KeyboardEvent} event the underlying dom event triggered for the gutter click
      */
    function onGutterClick(cm, line, gutter, event) {
        var opts = cm.state.foldGutter.options, pos = CodeMirror.Pos(line);
        if (gutter !== opts.gutter) { return; }
        var range;
        var _lineFolds = cm._lineFolds;
        if (cm.isFolded(line)) {
            if (event.altKey) { // unfold code including children
                range = _lineFolds[line];
                CodeMirror.commands.unfoldAll(cm, range.from.line, range.to.line);
            } else {
                cm.unfoldCode(line, {range: _lineFolds[line]});
            }
        } else {
            if (event.altKey) {
                range = CodeMirror.fold.auto(cm, pos);
                if (range) {
                    CodeMirror.commands.foldToLevel(cm, range.from.line, range.to.line);
                }
            } else {
                cm.foldCode(line);
            }
        }
    }

    /**
        Collapses all custom regions defined in the current editor
    */
    function collapseCustomRegions() {
        var editor = EditorManager.getFocusedEditor();
        if (editor) {
            var cm = editor._codeMirror, i = cm.firstLine();
            while (i < cm.lastLine()) {
                var range = cm.foldCode(i, {rangeFinder: regionFold});
                if (range) {
                    i = range.to.line;
                } else {
                    i++;
                }
            }
        }
    }

     /**
      * Collapses the code region nearest the current cursor position.
      * Nearest is found by searching from the current line and moving up the document until an
      * opening code-folding region is found.
      */
    function collapseCurrent() {
        var editor = EditorManager.getFocusedEditor();
        if (!editor) {
            return;
        }
        var cm = editor._codeMirror;
        var cursor = editor.getCursorPos(), i;
        // Move cursor up until a collapsible line is found
        for (i = cursor.line; i >= 0; i--) {
            if (cm.foldCode(i)) {
                editor.setCursorPos(i);
                return;
            }
        }
    }

    /**
      * Expands the code region at the current cursor position.
      */
    function expandCurrent() {
        var editor = EditorManager.getFocusedEditor();
        if (editor) {
            var cursor = editor.getCursorPos(), cm = editor._codeMirror;
            cm.unfoldCode(cursor.line);
        }
    }

    /**
        Collapses all foldable regions in the current document. Folding is done up to a level 'n'
        which is defined in the preferences. Levels refer to fold heirarchies e.g., for the following
        code fragment, the function is level 1, the if statement is level 2 and the forEach is level 3

            function sample() {
                if (debug) {
                    logMessages.forEach(function (m) {
                        console.debug(m);
                    });
                }
            }
    */
    function collapseAll() {
        var editor = EditorManager.getFocusedEditor();
        if (editor) {
            var cm = editor._codeMirror;
            CodeMirror.commands.foldToLevel(cm);
        }
    }

    /**
        Expands all folded regions in the current document
    */
    function expandAll() {
        var editor = EditorManager.getFocusedEditor();
        if (editor) {
            var cm = editor._codeMirror;
            CodeMirror.commands.unfoldAll(cm);
        }
    }

    /**
      * Initialises and creates the code-folding gutter.
      * @param {Editor} editor the editor on which to initialise the fold gutter
      */
    function createGutter(editor) {
        var cm = editor._codeMirror;
        var path = editor.document.file.fullPath, _lineFolds = prefs.getFolds(path);
        _lineFolds = _lineFolds || {};
        cm._lineFolds = _lineFolds;
        var gutters = cm.getOption("gutters").slice(0);

        // Reuse any existing fold gutter
        if (gutters.indexOf(GUTTER_NAME) < 0) {
            var lnIndex = gutters.indexOf("CodeMirror-linenumbers");
            $(editor.getRootElement()).addClass("folding-enabled");
            gutters.splice(lnIndex + 1, 0, GUTTER_NAME);
            cm.setOption("gutters",  gutters);
            cm.refresh();  // force recomputing gutter width - .folding-enabled class affects linenumbers gutter which has existing cached width
        }
        cm.setOption("foldGutter", {onGutterClick: onGutterClick});

        $(cm.getGutterElement()).on({
            mouseenter: function () {
                if (prefs.getSetting("hideUntilMouseover")) {
                    foldGutter.updateInViewport(cm);
                } else {
                    $(editor.getRootElement()).addClass("over-gutter");
                }
            },
            mouseleave: function () {
                if (prefs.getSetting("hideUntilMouseover")) {
                    foldGutter.clearGutter(cm);
                } else {
                    $(editor.getRootElement()).removeClass("over-gutter");
                }
            }
        });
    }

    /** Add gutter and restore saved expand/collapse state */
    function enableFoldingInEditor(editor) {
        if (editor._codeMirror.getOption("gutters").indexOf(GUTTER_NAME) === -1) {
            createGutter(editor);
            restoreLineFolds(editor);
        }
    }

     /**
      * When a brand new editor is seen, initialise fold-gutter and restore line folds in it. Save line folds in
      * departing editor in case it's getting closed.
      * @param {object} event the event object
      * @param {Editor} current the current editor
      * @param {Editor} previous the previous editor
      */
    function onActiveEditorChanged(event, current, previous) {
        if (current) {
            enableFoldingInEditor(current);
        }
        if (previous) {
            saveLineFolds(previous);
        }
    }

    /**
      * Saves the line folds in the current full editor before it is closed.
      */
    function saveBeforeClose() {
        // We've already saved all other open editors when they go active->inactive
        saveLineFolds(EditorManager.getActiveEditor());
    }


    function showSettingsDialog() {
        SettingsDialog.show(function () {
            var editor = EditorManager.getCurrentFullEditor();
            if (editor) {
                var cm = editor._codeMirror;
                if (prefs.getSetting("hideUntilMouseover")) {
                    foldGutter.clearGutter(cm);
                } else {
                    foldGutter.updateInViewport(cm);
                }
            }
        });
    }

    /**
        Initialise the extension
    */
    function init() {
        ExtensionUtils.loadStyleSheet(module, "main.less");
        foldCode.init();
        foldGutter.init();
        //register a global fold helper based on indentation folds
        CodeMirror.registerGlobalHelper("fold", "indent", function (mode, cm) {
            return prefs.getSetting("alwaysUseIndentFold");
        }, indentFold);

        CodeMirror.registerGlobalHelper("fold", "region", function (mode, cm) {
            return prefs.getSetting("enableRegionFolding");
        }, regionFold);

        CodeMirror.registerHelper("fold", "django", CodeMirror.helpers.fold.brace);
        CodeMirror.registerHelper("fold", "tornado", CodeMirror.helpers.fold.brace);

        $(EditorManager).on("activeEditorChange.BracketsCodeFolding", onActiveEditorChanged);
        $(DocumentManager).on("documentRefreshed.BracketsCodeFolding", function (event, doc) {
            if (prefs.getSetting("enabled")) {
                restoreLineFolds(doc._masterEditor);
            }
        });

        $(ProjectManager).on("beforeProjectClose.BracketsCodeFolding beforeAppClose.BracketsCodeFolding", saveBeforeClose);

        Menus.getMenu(Menus.AppMenuBar.VIEW_MENU).addMenuDivider();
        Menus.getMenu(Menus.AppMenuBar.VIEW_MENU).addMenuItem(CODE_FOLDING_SETTINGS);
        Menus.getMenu(Menus.AppMenuBar.VIEW_MENU).addMenuItem(COLLAPSE);
        Menus.getMenu(Menus.AppMenuBar.VIEW_MENU).addMenuItem(EXPAND);
        Menus.getMenu(Menus.AppMenuBar.VIEW_MENU).addMenuItem(COLLAPSE_ALL);
        Menus.getMenu(Menus.AppMenuBar.VIEW_MENU).addMenuItem(EXPAND_ALL);
        Menus.getMenu(Menus.AppMenuBar.VIEW_MENU).addMenuItem(COLLAPSE_CUSTOM_REGIONS);

        KeyBindingManager.addBinding(COLLAPSE, collapseKey);
        KeyBindingManager.addBinding(EXPAND, expandKey);
        KeyBindingManager.addBinding(COLLAPSE_ALL, collapseAllKey);
        KeyBindingManager.addBinding(EXPAND_ALL, expandAllKey);

        ///put region fold on top of the globals list so that it has priority over comments folds
        var finder = CodeMirror.fold._global.pop();
        if (finder) {
            CodeMirror.fold._global.unshift(finder);
        }

        var editor = EditorManager.getCurrentFullEditor();
        if (editor) {
            enableFoldingInEditor(editor);
            var cm = editor._codeMirror;
            if (prefs.getSetting("hideUntilMouseover")) {
                foldGutter.clearGutter(cm);
            } else {
                foldGutter.updateInViewport(cm);
            }
        }
    }

    AppInit.htmlReady(function () {
        CommandManager.register(Strings.CODE_FOLDING_SETTINGS + "...", CODE_FOLDING_SETTINGS, showSettingsDialog);
        CommandManager.register(Strings.COLLAPSE_ALL, COLLAPSE_ALL, collapseAll);
        CommandManager.register(Strings.EXPAND_ALL, EXPAND_ALL, expandAll);
        CommandManager.register(Strings.COLLAPSE_CUSTOM_REGIONS, COLLAPSE_CUSTOM_REGIONS, collapseCustomRegions);
        CommandManager.register(Strings.COLLAPSE_CURRENT, COLLAPSE, collapseCurrent);
        CommandManager.register(Strings.EXPAND_CURRENT, EXPAND, expandCurrent);
        if (prefs.getSetting("enabled") && !CodeMirror.fold.combine) {
            init();
        } else if (!prefs.getSetting("enabled")) {
            Menus.getMenu(Menus.AppMenuBar.VIEW_MENU).addMenuDivider();
            Menus.getMenu(Menus.AppMenuBar.VIEW_MENU).addMenuItem(CODE_FOLDING_SETTINGS);
        }
    });
});
